modules/ting_das/includes/ting_das.client.inc

```diff
@@ -0,0 +1,430 @@
+   *
+   * @return string|null
+   *   The article date.
+   */
+  public function getPublicationDateOfComponent() {
+    return $this->publicationDateOfComponent;
+  }
+
+  /**
+   * Set the article year.
+   *
+   * @param string $publicationYearOfComponent
+   *   Article year.
+   *
+   * @return TingDasCopyRequest
+   *   The current request object.
+   */
+  public function setPublicationYearOfComponent(string $publicationYearOfComponent) {
+    $this->publicationYearOfComponent = $publicationYearOfComponent;
+    return $this;
+  }
+
+  /**
+   * Get the article year.
+   *
+   * @return string|null
+   *   The article year.
+   */
+  public function getPublicationYearOfComponent() {
+    return $this->publicationYearOfComponent;
+  }
+
+  /**
+   * Set article issue number.
+   *
+   * @param string $issueOfComponent
+   *   Article issue number.
+   *
+   * @return TingDasCopyRequest
+   *   The current request object.
+   */
+  public function setIssueOfComponent(string $issueOfComponent) {
+    $this->issueOfComponent = $issueOfComponent;
+    return $this;
+  }
+
+  /**
+   * Get article issue number.
+   *
+   * @return string
+   *    The issue number.
+   */
+  public function getIssueOfComponent() {
+    return $this->issueOfComponent;
+  }
+
+  /**
+   * Set the the volume number.
+   *
+   * @param string $volumeOfComponent
+   *   The article volume number.
+   *
+   * @return TingDasCopyRequest
+   *   The current request object.
+   */
+  public function setVolumeOfComponent(string $volumeOfComponent) {
+    $this->volumeOfComponent = $volumeOfComponent;
+    return $this;
+  }
+
+  /**
+   * Get the article volume number.
+   *
+   * @return string|null
+   *   The article volume number.
+   */
+  public function getVolumeOfComponent() {
+    return $this->volumeOfComponent;
+  }
+
+  /**
+   * Set the page number in the publication.
+   *
+   * @param string $pagesOfComponent
+   *   The page number.
+   *
+   * @return TingDasCopyRequest
+   *   The current request object.
+   */
+  public function setPagesOfComponent(string $pagesOfComponent) {
+    $this->pagesOfComponent = $pagesOfComponent;
+    return $this;
+  }
+
+  /**
+   * Get the page number.
+   *
+   * @return string|null
+   *   The page number.
+   */
+  public function getPagesOfComponent() {
+    return $this->pagesOfComponent;
+  }
+
+  /**
+   * Send request to the service.
+   *
+   * @return bool
+   *   TRUE if request was send else FALSE.
+   */
+  public function send() {
+    $options = array(
+      'headers' => array(
+        'Content-Type' => 'application/xml; charset=UTF8',
+      ),
+      'body' => $this->buildMessage(),
+    );
+
+    try {
+      $client = new Client();
+      $response = $client->request('POST', $this->url, $options);
+
+      if ($response->getStatusCode() == 204) {
+        watchdog('ting_das', 'Service returned: !code', array('!code' => $response->getStatusCode()), WATCHDOG_ERROR);
+        return TRUE;
```

Comment:
Why do we check for status code 204 explicitly? It might be what the service currently returns if things go well but there is no handling of other success status codes even though they might be just as valid. Adding to this the method also returns `FALSE` in this situation indicating to the caller that the request was not sent.

[Guzzle throws an exception for all non-200 status codes and other errors](http://docs.guzzlephp.org/en/stable/quickstart.html) so such situations are well handled by the remaining code.

Regarding the logging also note:
- You have a nice non-Drupal specific class here. Consider injecting a PSR3 logger interface instead of calling `watchdog` directly and then use an adapter such as [`drupol/drupal7_psr3_watchdog`](https://packagist.org/packages/drupol/drupal7_psr3_watchdog).
- The current code sets log level `WATCHDOG_ERROR` for a successful call. That does not seem right.
- It can be extremely helpful to have an option which when enabled logs outgoing requests and responses. Consider adding that.

Reply:
It should have been != 204. But the service only have to stats 204 = accepted and 400 which it only throws when the XML message is not proper.

Reply:
Comment and and the only logger is the exception logger. I'am not sure if this request the effort to load yet another library into the site.